### ISSUE TITLE
LSTM training not equal when stateful=True/False units=1 samples=1 batch_size=1 epochs=1

### ISSUE THREAD

**Issue by mrhgit** [keras-team/keras#8198]
I'm trying to compare output when doing a stateful vs not LSTM, but I'm stuck on the first batch, because I can't get them to be equal and looking at the recurrent.py and tensorflow_backend.py isn't helping...

(transcribed the script, so hopefully it works)

### IMPORT
import keras
from keras.models import Sequential
from keras.layers import LSTM
import numpy as np

### BUILD MODELS
nonstateful_model = Sequential()
nonstateful_model.add(LSTM(1, input_shape=(1,1), stateful=False)
nonstateful_optimizer = keras.optimizers.SGD()
nonstateful_model.compile(loss='mean_squared_error', optimizer=nonstateful_optimizer)

stateful_model = Sequential()
stateful_model.add(LSTM(1, input_shape=(1,1), batch_size=1, stateful=True)
stateful_optimizer = keras.optimizers.SGD()
stateful_model.compile(loss='mean_squared_error', optimizer=stateful_optimizer)

### COPY WEIGHTS BETWEEN MODELS AND COMPARE TO VERIFY
stateful_model.set_weights(nonstateful_model.get_weights())
print(nonstateful_model.get_weights())
print(stateful_model.get_weights())

### TRAIN MODELS
x = np.array([[[1.]]],dtype='float32') # Noticed that the initial_states type will reflect x.dtype, but stateful states will be default float type
y = np.array([0.5],dtype='float32')
nonstateful_model.fit(x, y, epochs=1, batch_size=1, verbose=2, shuffle=False)
stateful_model.fit(x, y, epochs=1, batch_size=1, verbose=2, shuffle=False)

### COMPARE WEIGHTS AFTER TRAINING
print(nonstateful_model.get_weights())
print(stateful_model.get_weights())

- **2nd elements of the kernel arrays are different**
- recurrent kernels match perfectly
- **2nd elements of the bias arrays are different**

My goal is to be able to match kernels after training both models on N (x,y) pairs, where the nonstateful model is trained all at once and the stateful model is trained one pair at a time.

**Comment by kgrm** [comment]
Stateful models require a different formulation of the training data, since batches are assumed to correspond to time steps. See https://keras.io/layers/recurrent/ for details. And besides, there is no reason why two different models should be equal after training. The stateful parameter is a pretty fundamental change in the model.

**Comment by mrhgit** [comment]
I'm not doing multiple batches, so the statefulness should not affect anything at all.  According to the documentation, in this case, the weight updates _should definitely match_.  I purposefully chose timestep=1, epochs=1, batch_size=1 to make sure this was true.  **Note that all are indeed equivalent**, _**except for the kernel_f value and its corresponding bias_f**_.  This could very well be a bug, but first I'd like to simply pinpoint the reason why the values are different.
